### PR TITLE
Boost: Add Cloud Critical CSS generation status

### PR DESCRIPTION
**Pull request by karthikax**
Fixes 60-gh-Automattic/boost-hydra

#### Changes proposed in this Pull Request:

* Add Cloud Critical CSS generation status text below the module's main description.

#### Jetpack product discussion

NA

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Go to '..'
*

### CODE REVIEW

File: projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts [hunk 1]
```diff
@@ -2,7 +2,43 @@
...
-export function requestCloudCss(): Promise< string > {
-	return api.post( '/cloud-css/request-generate' );
+export async function requestCloudCss(): Promise< void > {
+	// Todo: Debounce request.
+	resetStatus();
+	await api.post( '/cloud-css/request-generate' );
+	pollStatus( true );
+}
+
+export async function pollCloudCssStatus() {
```
Comment (thingalon) [inline projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts:14]: I'm not sure about this logic.

If I understand it correctly, `pollCloudCssStatus()` decides how fast it should poll Critical CSS status, and is called under two circumstances:
1. When you mount the Cloud Critical CSS Module panel.
2. When a pending Cloud Critical CSS task comes back as completed.

The interval timer is only ever cancelled in the second situation - when a pending request comes back as completed.

I can't see
1. How the interval timer gets sped up when you click "regenerate",
2. What prevents multiple timers from running side-by-side in the future if we add more pages to the dashboard, and users navigate to sub-pages - and return to the main page, re-mounting the Module components.

IMHO we should keep one interval timer id (maybe as a module-level variable) which we check and clear before setting up any new interval timers. I also feel that passing `true` or `false` to `pollStatus` is a little hard to follow given `pollStatus` then has logic in it to manually check for `status.pending` in it anyway.

In my opinion, these two methods could be merged - or, if we want to split them up - have one method which returns an appropriate interval to poll for given a `status` object?

**Review by thingalon — CHANGES_REQUESTED** [review]
Seems to work, and looks good - though I am unsure of the logic around the timers. IMHO it could do with a tidy-up.

File: projects/plugins/boost/app/assets/src/js/pages/settings/sections/Modules.svelte [hunk 2]
```diff
@@ -54,11 +54,8 @@
 			/>
 		</p>
 
-		<div slot="meta">
-			<button type="button" class="components-button is-link" on:click={requestCloudCss}>
-				<RefreshIcon />
-				{__( 'Regenerate', 'jetpack-boost' )}
-			</button>
+		<div slot="meta" class="jb-feature-toggle__meta">
+			<GenerateCloudCss />
```
Comment (eAdnan007) [inline projects/plugins/boost/app/assets/src/js/pages/settings/sections/Modules.svelte:58]: I think the name `GenerateCloudCss` is a bit misleading. May be this should be something like `CloudCSSMeta`?

File: projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts (hunk 1 above)
Comment (eAdnan007) [reply projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts:14]: +1

**Review by eAdnan007 — CHANGES_REQUESTED** [review]
I second on Mark's point.

File: projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts (hunk 1 above)
Comment (karthikax) [reply projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts:14]: I agree with the tidy-up part, I will update and revert.

Meanwhile, a few info:
> `pollCloudCssStatus` is called under
> 2. When a pending Cloud Critical CSS task comes back as completed.

Not exactly. It is called when a generation request is completed so that it starts polling for updates given by the wpcom API.

> I can't see
> How the interval timer gets sped up when you click "regenerate",

Clicking `regenerate` calls `requestCloudCss` which resets the status and in turn changes the `pending` status to `true`. This makes the interval pick the shorter interval.

And, thanks for the review. Not clearing the interval in a loop is a major issue. I am fixing this.

File: projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts [hunk 3]
```diff
@@ -2,7 +2,43 @@
...
+
+	if ( statusIntervalId !== null ) {
+		clearInterval( statusIntervalId );
+	}
+
+	statusIntervalId = setInterval( async () => {
+		status = await api.get< CloudCssStatus >( '/cloud-css/status' );
+		updateStatus( status );
+
+		if ( ! status.pending && interval === shortInterval ) {
```
Comment (thingalon) [inline projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts:40]: This is much better. Though one final improvement suggestion, just to make the code super readable and easy to follow:

Add a function to this module called `pollIntervalForStatus( status )` which accepts a `status` object.

Then the start of this function can say:
```
const interval = pollIntervalForStatus( getStatus() );
```

And whenever you check if the interval should change, you can change this `if` to:
```
if ( interval !== pollIntervalForStatus( status ) )
```

That will cover both cases -- any time the interval wants to change because the status has changed, it will re-call `pollCloudCssStatus`.

What do you think? Just a nice change to make the code super easy to read.

**Review by thingalon — APPROVED** [review]
One suggestion, but not a blocker.

File: projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts (hunk 3 above)
Comment (karthikax) [reply projects/plugins/boost/app/assets/src/js/utils/cloud-css.ts:40]: Oh. Why didn't I think that! It works like a charm. Thanks @thingalon

**Review by thingalon — APPROVED** [review]
👍 Looks good! Thanks @karthikax
